### DISCUSSION TITLE
A 'rechuandao' is not a registered object

### DISCUSSION THREAD

**Discussion by Ating24** [idaholab/moose#23935] [Q&A General] [answered]
I am a newcomer to MOOSE and have written an app on heat conduction in a steady state, but the following error occurred:
```
*** ERROR ***
A 'rechuandao' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

Stack frames: 20
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: void mooseError<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&)
3: Factory::reportUnregisteredError(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
4: Factory::getValidParams(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
5: MooseObjectAction::MooseObjectAction(InputParameters const&)
6: AddKernelAction::AddKernelAction(InputParameters const&)
7: /home/liu/projects/moose/framework/libmoose-opt.so.0(+0x1419918) [0x7fd98860e918]
8: ActionFactory::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters&)
9: Parser::walkRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, hit::Node*)
10: Parser::walk(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, hit::Node*)
11: hit::Node::walk(hit::Walker*, hit::NodeType, hit::TraversalOrder)
12: hit::Node::walk(hit::Walker*, hit::NodeType, hit::TraversalOrder)
13: hit::Node::walk(hit::Walker*, hit::NodeType, hit::TraversalOrder)
14: Parser::parse(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)
15: MooseApp::setupOptions()
16: MooseApp::run()
17: main
18: __libc_start_main
19: ./hc-opt(+0x23ef) [0x55915a9803ef]
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
``` 

This is the header file：
/
```c++
#pragma once

#include "Diffusion.h"


class rechuandao : public Diffusion
{
public:
  static InputParameters validParams();

  rechuandao(const InputParameters & parameters);

protected:
  
  virtual Real computeQpResidual() override;

  const Real _k;
};



```

This is the source file：
/ ```c++
#include "rechuandao.h"

registerMooseObject("hcAPP", rechuandao);

InputParameters
rechuandao::validParams()
{
    InputParameters params = Diffusion::validParams();
    
    params.addRequiredParam<Real>("k", "热导率");
    
    return params;
}

rechuandao::rechuandao(const InputParameters & parameters)
  : Diffusion(parameters),
   _k(getParam<Real>("k"))
    
{
}

Real
rechuandao::computeQpResidual()
{
    return (_k * _grad_u[_qp] * _grad_test[_i][_qp]);
}

``` 



This is an i file：
```
[Mesh]
  type = GeneratedMesh # Can generate simple lines, rectangles and rectangular prisms
  dim = 2 # Dimension of the mesh
  nx = 10 # Number of elements in the x direction
  ny = 10 # Number of elements in the y direction
  xmax = 0.01 # Length of test chamber
  ymax = 0.01 # Test chamber radius
[]
[Variables]
  [./temperature]
    order = FIRST
    family = LAGRANGE
  [../]
[]
[Kernels]
  [diffusion]
    type = rechuandao 
    variable = temperature 
    k=1
  []
[]
[BCs]
  [./bottom] 
    type = DirichletBC
    variable = temperature
    boundary = 'bottom' # This must match a named boundary in the mesh file
    value = 300
  [../]

  [./top] 
    type = DirichletBC
    variable = temperature
    boundary = 'top' # This must match a named boundary in the mesh file
    value = 1000
  [../]
[]
[Executioner]
  type = Steady # Steady state problem
  solve_type = 'PJFNK'

[]
[Outputs]
  exodus = true # Output Exodus format
[]

```

**Comment by GiudGiud** [comment, marked as answer]
registerMooseObject("hcAPP", rechuandao);

hcAPP is the name of the app in the base/hcAPP.C file too?
It must match exactly

**Reply by Ating24** [reply]
Yes, the created app is called hc.

**Reply by GiudGiud** [reply]
but hcAPP or hcApp ?

**Reply by Ating24** [reply]
Okay, just change the hcApp to hcApp, thank you
